### PR TITLE
Adding various operations and examples for Budget APIs

### PR DESCRIPTION
**Pull request by asarkar84**
Adding various operations and examples for Budget APIs

1. Operation to Get the List of Budgets for a subscription.
2. Operation to Get the specific budget by name for a subscription.
3. Operation to Create or Update a budget.
4. Operation to Delete an existing budget.
5. Examples for Create, Update, Delete, Get and List budgets.

This checklist is used to make sure that common issues in a pull request are addressed. This will expedite the process of getting your pull request merged and avoid extra work on your part to fix issues discovered during the review process. 

### PR information
- [x] The title of the PR is clear and informative.
- [x] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For information on cleaning up the commits in your pull request, [see this page](https://github.com/Azure/azure-powershell/blob/dev/documentation/cleaning-up-commits.md).
- [x] Except for special cases involving multiple contributors, the PR is started from a fork of the main repository, not a branch.
- [ ] If applicable, the PR references the bug/issue that it fixes.
- [x] Swagger files are correctly named (e.g. the `api-version` in the path should match the `api-version` in the spec).

### Quality of Swagger
- [x] I have read the [contribution guidelines](https://github.com/Azure/azure-rest-api-specs/blob/master/.github/CONTRIBUTING.md).
- [x] My spec meets the review criteria:
  - [x] The spec conforms to the [Swagger 2.0 specification](https://github.com/OAI/OpenAPI-Specification/blob/master/versions/2.0.md).
  - [x] The spec follows the guidelines described in the [Swagger checklist](../documentation/swagger-checklist.md) document.
  - [x] [Validation tools](https://github.com/Azure/azure-rest-api-specs/blob/master/documentation/swagger-checklist.md#validation-tools-for-swagger-checklist) were run on swagger spec(s) and have all been fixed in this PR.

### CODE REVIEW

**Review by fearthecowboy — CHANGES_REQUESTED** [review]
Do not move the file common-types file; please reference it where it is.

ie, do not rename or copy the file :

/specification/common-types/resource-management/v1/types.json

File: specification/consumption/resource-manager/Microsoft.Consumption/2017-12-30-preview/consumption.json [hunk 1]
```diff
@@ -0,0 +1,456 @@
+{
+  "swagger": "2.0",
+  "info": {
+    "version": "2017-12-30-preview",
+    "title": "ConsumptionManagementClient",
```
Comment (fearthecowboy) [inline specification/consumption/resource-manager/Microsoft.Consumption/2017-12-30-preview/consumption.json]: `title` should just be `"Consumption Management"`

File: specification/consumption/resource-manager/Microsoft.Consumption/2017-12-30-preview/consumption.json [hunk 2]
```diff
@@ -0,0 +1,456 @@
...
+          "nextLinkName": "null"
+        }
+      }
+    },
+    "/subscriptions/{subscriptionId}/providers/Microsoft.Consumption/budgets/{budgetName}": {
+      "get": {
+        "tags": [
+          "Budget"
+        ],
+        "operationId": "Budget_Get",
```
Comment (fearthecowboy) [inline specification/consumption/resource-manager/Microsoft.Consumption/2017-12-30-preview/consumption.json]: the prefix on `operationId` is generally plural; so should be `budgets_...`

File: specification/consumption/resource-manager/Microsoft.Consumption/2017-12-30-preview/consumption.json [hunk 3]
```diff
@@ -0,0 +1,456 @@
...
+      "description": "Error response indicates that the service is not able to process the incoming request. The reason is provided in the error message.",
+      "type": "object",
+      "properties": {
+        "error": {
+          "description": "The details of the error.",
+          "$ref": "#/definitions/ErrorDetails"
+        }
+      }
+    },
+    "Resource": {
```
Comment (fearthecowboy) [inline specification/consumption/resource-manager/Microsoft.Consumption/2017-12-30-preview/consumption.json]: `Resource` should be referenced from  the `common-types` file, not copied

File: specification/consumption/resource-manager/Microsoft.Consumption/2017-12-30-preview/consumption.json [hunk 4]
```diff
@@ -0,0 +1,456 @@
...
+          },
+          "default": {
+            "description": "Error response describing why the operation failed.",
+            "schema": {
+              "$ref": "#/definitions/ErrorResponse"
+            }
+          }
+        },
+        "x-ms-pageable": {
+          "nextLinkName": "null"
```
Comment (fearthecowboy) [inline specification/consumption/resource-manager/Microsoft.Consumption/2017-12-30-preview/consumption.json]: So your `nextLink` Property Name is `"null"` ?

**Review by fearthecowboy — CHANGES_REQUESTED** [review]
I've added a few bits of feedback; I'll add more in a bit...

**Comment by fearthecowboy** [comment]
Adding @ravbhatnagar  for ARM review of new APIS

File: specification/consumption/resource-manager/Microsoft.Consumption/2017-12-30-preview/consumption.json (hunk 1 above)
Comment (sandeepnl) [reply specification/consumption/resource-manager/Microsoft.Consumption/2017-12-30-preview/consumption.json]: I think last time when it was changed @lmazuel Suggested it this way due to the requirement on the SDK to generate the client name correctly.

Comment (lmazuel) [reply specification/consumption/resource-manager/Microsoft.Consumption/2017-12-30-preview/consumption.json]: @fearthecowboy Again, this is a C# behavior. Please move this "auto Client suffixing" to the modeler if you wish this behavior for everybody. In the meantime, changing that will break Python (at least, but probably not only)

**Comment by asarkar84** [comment]
I could not find the eTag property for ProxyResource in common-types. Shall I go ahead and add the eTag property there and use it in this file.

**Comment by fearthecowboy** [comment]
> I could not find the eTag property for ProxyResource in common-types. Shall I go ahead and add the eTag property there and refer it in this file?

I just had this same discovery with another RP this morning.

For today, can you just add the etag property into your resource model(s) ; I think we have to have a second version of some of these (for older RPs that don't have all the fields).

We'll circle back once I can go bang on a few doors.

File: specification/consumption/resource-manager/Microsoft.Consumption/2017-12-30-preview/consumption.json (hunk 1 above)
Comment (fearthecowboy) [reply specification/consumption/resource-manager/Microsoft.Consumption/2017-12-30-preview/consumption.json]: @lmazuel Right!  @sandeepnl

**Review by sandeepnl — CHANGES_REQUESTED** [review]
Let us rename this to "Budget-PREVIEW" as we want this name to show up in the Rest Documentation.

File: specification/consumption/resource-manager/Microsoft.Consumption/2017-12-30-preview/consumption.json (hunk 1 above)
Comment (asarkar84) [reply specification/consumption/resource-manager/Microsoft.Consumption/2017-12-30-preview/consumption.json]: Sure! We will leave it as is. Thanks!

File: specification/consumption/resource-manager/Microsoft.Consumption/2017-12-30-preview/consumption.json [hunk 5]
```diff
@@ -0,0 +1,471 @@
...
+      "scopes": {
+        "user_impersonation": "impersonate your user account"
+      }
+    }
+  },
+  "paths": {
+    "/subscriptions/{subscriptionId}/providers/Microsoft.Consumption/budgets": {
+      "get": {
+        "tags": [
+          "Budgets"
```
[the remark this answers is not in the document]
Comment (asarkar84) [reply specification/consumption/resource-manager/Microsoft.Consumption/2017-12-30-preview/consumption.json]: Sure! I am going to rename the tag to Budgets-Preview @fearthecowboy Please let me know if this is OK.

**Comment by fearthecowboy** [comment]
You have some errors in validation: 
https://travis-ci.org/Azure/azure-rest-api-specs/jobs/316251192
https://travis-ci.org/Azure/azure-rest-api-specs/jobs/316251191

Notably; 
  `eTag` vs `etag`

**Comment by fearthecowboy** [comment]
Rather than have `ProxyResource` , just add an eTag property at https://github.com/asarkar84/azure-rest-api-specs/blob/79f3e9378bc86d6dc7cb338d983cc5db3a202e96/specification/consumption/resource-manager/Microsoft.Consumption/2017-12-30-preview/consumption.json#L260  and have `Budget` reference `Resource` from the common file?

**Review by sandeepnl — APPROVED** [review]

**Review by fearthecowboy — APPROVED** [review]

**Pull request merged**